### ISSUE TITLE
Suggestion for FindImagesWithPossibleResection

### ISSUE THREAD

**Issue by rdiazgar** [openMVG/openMVG#312]
Hi,

While performing an incremental reconstruction, I find that function FindImagesWithPossibleResection is particularly slow, especially for large datasets. I was wondering if we can take advantage of OpenMP in this situation and came up with a workaround that seems to be working. I'm currently running this modified version, and while the reconstruction has not ended yet, temporary results are looking good while speed has increased considerably.

In the loop that looks for the next images to add, the code loops an std::set of image indices not yet added. While sets are annoying to use in OpenMP, we can turn this set into a temporary vector and do a parallel loop like this:

```
/*************************************/
// Copy set into vector and perform parallel loop
std::vector<size_t> idList(_set_remainingImageId.begin(), _set_remainingImageId.end());
std::vector< std::pair<size_t, size_t> > vec_putative; // ImageId, NbPutativeCommonPoint

#ifdef USE_OPENMP
  #pragma omp parallel for schedule(dynamic, 1) 
#endif
  for (size_t i = 0; i < idList.size(); ++i)
//  for (std::set<size_t>::const_iterator iter = _set_remainingImageId.begin();
//        iter != _set_remainingImageId.end(); ++iter)
  {
    //const size_t imageIndex = *iter;
    const size_t imageIndex = idList[i];
... 
... 
...
#ifdef USE_OPENMP
  #pragma omp critical
#endif
      {
      // make sure the pair is added once at a time.
      vec_putative.push_back( make_pair(imageIndex, vec_trackIdForResection.size()));
      }
    }
  }
/***************************************/
```

**Comment by pmoulon** [comment]
Hi @rdiazgar ,
Thanks for the suggestion.

The incremental SfM pipeline is like a demonstrator and a lot of steps can be tweaked or run in parallel.

In any incremental pipeline you must have a rule to find the next set of image for pose estimation:
- next candidate image are chosen dynamically from 3D-2D visibility (openMVG, Bundler rule).
- candidate list can be computed at the start of the process by using a MST from the pair seed. Runtime to choose the next image can be constant in some case, but perhaps less optimal (visualSfM rule).

In OpenMVG case, since outliers track observations are removed during the process, I have preferred to list TrackId per view at each round.

Note: To speed up the actual algorithm you can do the following (it will require more storage in RAM)
 - you could precompute the trackId per view in an initialization set (when tracks have been computed)  and use them at each round.  

Regarding using the set in an OpenMP parallel section, you can try the following by using the section openMP keyword:
(note that the code is made for the last OpenMVG version, develop or master branch: USE_OPENMP -> OPENMVG_USE_OPENMP)

```
#ifdef OPENMVG_USE_OPENMP
  #pragma omp parallel
#endif
  for (std::set<size_t>::const_iterator iter = _set_remainingViewId.begin();
        iter != _set_remainingViewId.end(); ++iter)
  {
#ifdef OPENMVG_USE_OPENMP
  #pragma omp single nowait
#endif
    {
      const size_t viewId = *iter;

      // Compute 2D - 3D possible content
      openMVG::tracks::STLMAPTracks map_tracksCommon;
      std::set<size_t> set_viewId;
      set_viewId.insert(viewId);
      tracks::TracksUtilsMap::GetTracksInImages(set_viewId, _map_tracks, map_tracksCommon);

      if (! map_tracksCommon.empty())
      {
        std::set<size_t> set_tracksIds;
        tracks::TracksUtilsMap::GetTracksIdVector(map_tracksCommon, &set_tracksIds);

        // Count the common possible putative point
        //  with the already 3D reconstructed trackId
        std::vector<size_t> vec_trackIdForResection;
        set_intersection(set_tracksIds.begin(), set_tracksIds.end(),
          reconstructed_trackId.begin(),
          reconstructed_trackId.end(),
          std::back_inserter(vec_trackIdForResection));

#ifdef OPENMVG_USE_OPENMP
        #pragma omp critical
#endif
        {
          vec_putative.push_back( make_pair(viewId, vec_trackIdForResection.size()));
        }
      }
    }
  }
```

I have tried on a medium sized dataset and on a machine with 8 cores, I have a ~= 66% speed-up in this function.

So using MT in this section make the process more memory hungry, but faster ;-).
I note that point for the next OpenMVG release candidate.

**Comment by rdiazgar** [comment]
Hi Pierre,

Happy that I could help. I'm not very expert on openMP and came up with the solution by googling the openMP basics. Your solution is more elegant and does not need any extra variables, which is good.

Kind regards,
Raúl

**Comment by pmoulon** [comment]
Included in OpenMVG v0.8.1.
Thanks for your suggestion @rdiazgar
